internal/engine/experiment/quicping/quicping.go

```diff
@@ -0,0 +1,443 @@
+	err      error
+}
+
+// sender sends a ping requests to the target hosts every second
+func (m *Measurer) sender(
+	ctx context.Context,
+	pconn model.UDPLikeConn,
+	destAddr *net.UDPAddr,
+	out chan<- requestInfo,
+	sess model.ExperimentSession,
+) {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	i := 0
+	for i < int(m.config.repetitions()) {
+		select {
+		case <-ctx.Done():
+			return // user aborted or timeout expired
+
+		case sendTime := <-ticker.C:
+			packet, dstID, srcID := buildPacket()     // build QUIC Initial packet
+			_, err := pconn.WriteTo(packet, destAddr) // send Initial packet
+
+			sess.Logger().Infof("PING %s", destAddr)
+
+			// propagate send information, including errors
+			out <- requestInfo{raw: packet, t: sendTime, dstID: hex.EncodeToString(dstID), srcID: hex.EncodeToString(srcID), err: err}
+			i += 1
+
+		default:
+		}
+	}
+}
+
+// receiver receives incoming server responses and
+// dissects the payload of the version negotiation response
+func (m *Measurer) receiver(
+	ctx context.Context,
+	pconn model.UDPLikeConn,
+	out chan<- responseInfo,
+	sess model.ExperimentSession,
+) {
+	for ctx.Err() == nil {
+		// read (timeout was set in Run)
+		buffer := make([]byte, 1024)
+		n, addr, err := pconn.ReadFrom(buffer)
+		respTime := time.Now()
+		if err != nil {
+			// stop if the connection is already closed
+			if errors.Is(err, net.ErrClosed) {
+				break
+			}
+			// store read failures and continue receiving
+			out <- responseInfo{t: respTime, err: err}
+			continue
+		}
+		resp := buffer[:n]
+
+		// dissect server response
+		supportedVersions, dst, err := m.DissectVersionNegotiation(resp)
+		if err != nil {
+			// the response was likely not the expected version negotiation response, so ignore it
+			sess.Logger().Infof(fmt.Sprintf("response dissection failed: %s", err))
+			continue
+		}
+		// propagate receive information
+		out <- responseInfo{raw: resp, t: respTime, dstID: hex.EncodeToString(dst), versions: supportedVersions}
+
+		sess.Logger().Infof("PING got response from %s", addr)
+	}
+}
+
+// Run implements ExperimentMeasurer.Run.
+func (m *Measurer) Run(
+	ctx context.Context,
+	sess model.ExperimentSession,
+	measurement *model.Measurement,
+	callbacks model.ExperimentCallbacks,
+) error {
+	host := string(measurement.Input)
+	// allow URL input
+	if u, err := url.ParseRequestURI(host); err == nil {
+		host = u.Host
+	}
+	service := net.JoinHostPort(host, m.config.port())
+	udpAddr, err := net.ResolveUDPAddr("udp4", service)
+	if err != nil {
+		return err
+	}
+	rep := m.config.repetitions()
+	tk := &TestKeys{
+		Domain:      host,
+		Repetitions: rep,
+	}
+	measurement.TestKeys = tk
+
+	// create UDP socket
+	pconn, err := m.config.networkLibrary().ListenUDP("udp", &net.UDPAddr{})
+	if err != nil {
+		return err
+	}
+	defer pconn.Close()
+
+	// set context and read timeouts
+	deadline := time.Duration(rep*5) * time.Second
+	pconn.SetReadDeadline(time.Now().Add(deadline))
```

Comment:
I don't think it could happen in practice that we're blocked on `Write`. Though, to be on the safe side, I'd really  use `SetDeadline` here rather than `SetReadDeadline` so we know the `pconn` will eventually timeout.

(Maybe it's not needed but it's more robust to future refactorings.)

```suggestion
	pconn.SetDeadline(time.Now().Add(deadline))
```